### PR TITLE
[1.13] Tweak fluids to be registered like blocks and items

### PR DESCRIPTION
**Pull request by Madsthunder**
Original pr: #3664 

This pr moves the registry related methods to the `ForgeRegistry` system while trying to maintain backwards compatibility for 1.11.

As always, with breaking changes, there will need to be some things changed on the modders side. This includes:
- Having to register a fluid through either `GameRegistry.register`, `ForgeRegistries.FLUIDS.register`, or subscribing a `RegistryEvent.Register<Fluid>` to the `EventBus`.
- Adding the corresponding modid to the beginning of a custom fluid model (See the bottom of the changed files).

### CODE REVIEW

File: src/main/java/net/minecraftforge/common/ForgeModContainer.java [hunk 1]
```diff
@@ -503,15 +504,32 @@ public NBTTagCompound getDataForWriting(SaveHandler handler, WorldInfo info)
...
         return forgeData;
     }
 
     @Override
     public void readData(SaveHandler handler, WorldInfo info, Map<String, NBTBase> propertyMap, NBTTagCompound tag)
     {
         DimensionManager.loadDimensionDataMap(tag.hasKey("DimensionData") ? tag.getCompoundTag("DimensionData") : null);
-        FluidRegistry.loadFluidDefaults(tag);
+        if (tag.hasKey("DefaultFluidList", 9)) //Load Pre-1.12 Fluids
+    	{
```
Comment (liach) [inline src/main/java/net/minecraftforge/common/ForgeModContainer.java]: tab!

**Comment by mezz** [comment]
Please resolve the merge conflicts for this PR.

**Comment by Madsthunder** [comment]
Any breaking changes window for 1.12.1 anytime soon?

**Comment by LexManos** [comment]
1.12.1 is to small of a update to do breaking changes.

File: src/main/java/net/minecraftforge/registries/GameData.java [hunk 2]
```diff
@@ -98,11 +117,13 @@
     private static final int MAX_POTIONTYPE_ID = Integer.MAX_VALUE >> 5; // Int (SPacketEffect)
     private static final int MAX_ENCHANTMENT_ID = Short.MAX_VALUE - 1; // Short - serialized as a short in ItemStack NBTs.
     private static final int MAX_ENTITY_ID = Integer.MAX_VALUE >> 5; // Varint (SPacketSpawnMob)
+    private static final int MAX_FLUID_ID = 4096; //There are only 4096 possible blocks
```
Comment (liach) [inline src/main/java/net/minecraftforge/registries/GameData.java]: or 4095 like blocks?

File: src/main/java/net/minecraftforge/registries/GameData.java [hunk 3]
```diff
@@ -272,7 +300,31 @@ void clear()
...
+        public void onAdd(IForgeRegistryInternal<Fluid> owner, RegistryManager stage, int id, Fluid fluid, @Nullable Fluid oldFluid)
+        {
+            if(fluid.canBePlacedInWorld())
+                owner.getSlaveMap(FLUID_TO_BLOCK, BiMap.class).put(fluid, fluid.getBlock());
+        }
+
+        @Override
+        public void onCreate(IForgeRegistryInternal<Fluid> owner, RegistryManager stage)
+        {
+            owner.setSlaveMap(FLUID_TO_BLOCK, HashBiMap.<Fluid, Block>create());
```
Comment (liach) [inline src/main/java/net/minecraftforge/registries/GameData.java]: extra generics

File: src/main/java/net/minecraftforge/registries/GameData.java (hunk 2 above)
Comment (bs2609) [reply src/main/java/net/minecraftforge/registries/GameData.java]: Just use `MAX_BLOCK_ID`, it's already declared?

File: src/main/java/net/minecraftforge/registries/GameData.java (hunk 3 above)
Comment (liach) [reply src/main/java/net/minecraftforge/registries/GameData.java]: Not really, that is not type specific

**Review by liach — APPROVED** [review]

File: src/main/java/net/minecraftforge/registries/GameData.java (hunk 2 above)
Comment (kashike) [reply src/main/java/net/minecraftforge/registries/GameData.java]: No, this should really be its own constant like it is now. Changing the maximum block id in the future should not magically change the maximum fluid id.

Comment (bs2609) [reply src/main/java/net/minecraftforge/registries/GameData.java]: What other reason is there to limit the number of fluids though? If it should be its own constant, then it shouldn't be defined as 4096.

**Comment by liach** [comment]
@Madsthunder Can you make it compile and fix the merge conflicts?

File: src/main/java/net/minecraftforge/fluids/FluidRegistry.java [hunk 4]
```diff
@@ -301,168 +111,16 @@ public static boolean addBucketForFluid(Fluid fluid)
...
-        String name = masterFluidReference.inverse().get(key);
-        if (Strings.isNullOrEmpty(name)) {
-            FMLLog.log.error("The fluid registry is corrupted. A fluid {} {} is not properly registered. The mod that registered this is broken", key.getClass().getName(), key.getName());
-            throw new IllegalStateException("The fluid registry is corrupted");
-        }
-        return name;
-    }
-
-    @Nullable
-    public static String getModId(@Nullable FluidStack fluidStack)
```
Comment (liach) [inline src/main/java/net/minecraftforge/fluids/FluidRegistry.java:360]: This method is still called in `UniversalBucket.java`, causing build failures. Please replace it.

**Review by liach — CHANGES_REQUESTED** [review]

**Comment by Madsthunder** [comment]
As a sidenote, I will go back through the changes in this pull and reapply it on top of forge 1.13 once it comes around.

File: src/main/java/net/minecraftforge/fluids/FluidDictionary.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        {
+            NAME_TO_FLUIDS.put(name, NonNullList.<Fluid>withSize(1, fluid));
+        }
+    }
+    
+    public static boolean hasName(Fluid fluid, String name)
+    {
+    	if (FLUID_TO_NAMES.containsKey(fluid))
+    	{
+    		return FLUID_TO_NAMES.get(fluid).contains(name.toLowerCase(Locale.ENGLISH));
```
Comment (liach) [inline src/main/java/net/minecraftforge/fluids/FluidDictionary.java]: tab

File: src/main/java/net/minecraftforge/registries/GameData.java [hunk 6]
```diff
@@ -104,11 +123,13 @@
     private static final int MAX_POTIONTYPE_ID = Integer.MAX_VALUE >> 5; // Int (SPacketEffect)
     private static final int MAX_ENCHANTMENT_ID = Short.MAX_VALUE - 1; // Short - serialized as a short in ItemStack NBTs.
     private static final int MAX_ENTITY_ID = Integer.MAX_VALUE >> 5; // Varint (SPacketSpawnMob)
+    private static final int MAX_FLUID_ID = 4095; //There are only 4095 possible blocks
```
Comment (liach) [inline src/main/java/net/minecraftforge/registries/GameData.java:109]: Now 1.13 removed upper limit on block ids. You might seek to change this.

**Comment by HenryLoenwind** [comment]
And still the question is: Do we really want to introduce dozens of duplicate fluids? Player will have to fight with this, just like they have to deal with ways to de-duplicate their ores and ingots currently. Mod authors will also have to do so, making extra work for everyone who has any contact with fluids. Results will be very mixed---some modder may decide to auto-merge any two fluids that are registered for the same name, others will let fluids pass through as they are. And half of them will forget about fluid nbt and trash fluids from certain mods.

The current fluid system has no major drawbacks (for players or modders). Every mod author can decide if they want to use a common fluidname like "milk" or "creosote", or if they want to prefix the name with their modid to make it truly unique.

If this is just about using the registry system to handle things in the background, there would be a way to do this and still keep the existing semantics and FluidRegistry API: Keep the fluidname as is and internally set a registry name of "forge:"+fluidname. Add a modid field to the fluid if you want to track who registered it.

**Comment by liach** [comment]
@HenryLoenwind By that time, we can use unification tags! :scream: (1.13 new feature)

**Comment by mezz** [comment]
@HenryLoenwind the current registry has issues, see #1798 for one potential issue. It would also be nice to have everything consistent, so that fluids get the same features as in other registries.
A unification registry like in https://github.com/MinecraftForge/MinecraftForge/pull/4457 would be able to unify fluids together.

**Comment by liach** [comment]
This should not be stale.

**Comment by Madsthunder** [comment]
Just to be clear, there will be NO breaking changes window until 1.13 is released a few months from now, correct?

**Comment by AlexIIL** [comment]
@Madsthunder That's been the rule so far, as otherwise new versions forge would break existing mods. I think you can safely assume that there is little to do until 1.13.

(This is something the forge team has been very strict on so far, even with small changes - and this is a very large change, which will immediately break all mods that use fluids).

**Comment by Madsthunder** [comment]
Updated to work with the latest version of forge

**Comment by mezz** [comment]
@Madsthunder please update to resolve conflicts.

**Comment by LexManos** [comment]
Fluids are a vanilla concept now, and a full registry. This isn't needed.

**Pull request closed without merging**